File: solutions/1886-minimum-limit-of-balls-in-a-bag/minimum-limit-of-balls-in-a-bag.swift
class Solution {
    private func canDivide(_ nums: [Int], _ maxBalls: Int, _ maxOperations: Int) -> Bool {
        var ops = 0
        for n in nums {
            ops += (n + maxBalls - 1) / maxBalls - 1
            if ops > maxOperations { return false }
        }
        return true
    }
    
    func minimumSize(_ nums: [Int], _ maxOperations: Int) -> Int {
        var left = 1
        var right = nums.max()!
        var res = right
        
        while left < right {
            let mid = left + (right - left) / 2
            if canDivide(nums, mid, maxOperations) {
                right = mid
                res = right
            } else {
                left = mid + 1
            }
        }
        return res
    }
}

File: solutions/1886-minimum-limit-of-balls-in-a-bag/minimum-limit-of-balls-in-a-bag.rs
impl Solution {
    fn can_divide(nums: &Vec<i32>, max_balls: i32, max_operations: i32) -> bool {
        let mut ops = 0;
        for &n in nums {
            ops += (n + max_balls - 1) / max_balls - 1;
            if ops > max_operations {
                return false;
            }
        }
        true
    }
    
    pub fn minimum_size(nums: Vec<i32>, max_operations: i32) -> i32 {
        let mut left = 1;
        let mut right = *nums.iter().max().unwrap();
        let mut res = right;
        
        while left < right {
            let mid = left + (right - left) / 2;
            if Self::can_divide(&nums, mid, max_operations) {
                right = mid;
                res = right;
            } else {
                left = mid + 1;
            }
        }
        res
    }
}